### PR TITLE
enhancement: Add LEP 'Refactor restore for rebuild enabling'

### PR DESCRIPTION
**Pull request by shuo-wu**
#1279

### CODE REVIEW

File: enhancements/20200721-refactor-restore-for-rebuild-enabling.md [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+    - In order to skip the file syncing and snapshotting and directly do restore, the rebuilding related API should be updated, which will lead to API version bumps.
+- For DR volumes, the last restored backup will be different among all replicas and some replicas may be in restore progress when the rebuilding is triggered. Hence, Longhorn manager should not specify the last backup in the restore command. In other words, engines and replicas should be intelligent enough to decide if they need to do restore and which kind of restore they need to launch.
+    - Actually replicas have all information about the restore status and they can decide if they need incremental restore or full restore by themself. Specifying the last backup in the restore command is redundant.
+    - Longhorn manager only needs to tell the replicas what is the lastest backup they should restore.
```
Comment (yasker) [inline enhancements/20200721-refactor-restore-for-rebuild-enabling.md:23]: typo `latest` not `lastest`.

File: enhancements/20200721-refactor-restore-for-rebuild-enabling.md [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+14. Wait for the DR volume activated. Then launch a pod for the activated DR volume.
+15. Verify the restored data of the activated DR volume.
+16. Write more data to the activated DR volume. Then verify all replicas are still running.
+17. Crash one random replica of the activated DR volume. Then verify a new replica will be rebuild and the volume still works fine.
+
+### Upgrade strategy
+Since there is modification for controller gRPC API, the live upgrade is not supported.
+
+## Note
+It's possible that the restore/DR volume rebuilding somehow get stuck, or users has no time to wait for the restore/DR volume rebuilding complete. We need to provide a way that users can use the volume as soon as possible. This enhancement is trakced in https://github.com/longhorn/longhorn/issues/1512.
```
Comment (yasker) [inline enhancements/20200721-refactor-restore-for-rebuild-enabling.md:159]: `tracked`

File: enhancements/20200721-refactor-restore-for-rebuild-enabling.md [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+    - The normal rebuilding (file syncing) workflow implicitly considers that all existing snapshots won't change and newer data during the rebuilding will be written into volume head. But for restore/DR volumes, new data writing is directly handled by replica (sync agent server) and it will be written to underlying snapshots rather than volume heads. As a result, the normal rebuilding logic doesn't fit restore/DR volumes.  
+    - In order to skip the file syncing and snapshotting and directly do restore, the rebuilding related API should be updated, which will lead to API version bumps.
+- For DR volumes, the last restored backup will be different among all replicas and some replicas may be in restore progress when the rebuilding is triggered. Hence, Longhorn manager should not specify the last backup in the restore command. In other words, engines and replicas should be intelligent enough to decide if they need to do restore and which kind of restore they need to launch.
```
Comment (joshimoo) [inline enhancements/20200721-refactor-restore-for-rebuild-enabling.md:21]: > Longhorn manager should not specify the last backup in the restore command
> Longhorn manager only needs to tell the replicas what is the lastest backup they should restore.

Can you explain the difference.
Does last above refer to the previous backup?
While latest below refers to the newest backup? (in terms timeline)

File: enhancements/20200721-refactor-restore-for-rebuild-enabling.md [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+    - Longhorn manager still need to know what is the last restored backup of all replicas, since it relies on it to determine if the restore/DR volume is available/can be activated.
+- Longhorn should wait for rebuild complete and check restore status before auto detachment.
+    - Otherwise, the restore volume will be automatically detached when the rebuild is in progress then the rebuild is meaningless in this case.  
+
+### User Stories
+
+#### Replica crashes when a restore/DR volume is in restore progress
+Before, the restore volume keeps state `Degraded` if there is replica crashing. And the volume will finally become `Faulted` if replicas are crashed one by one during restoring.
+
+After, the restore volume will start replica rebuilding automatically then be back to state `Healthy` if there is replica crashing. The volume is available as long as all replicas are not crashed at the same time. And volume will finish activation/auto-detachment after the rebuild is done.
```
Comment (joshimoo) [inline enhancements/20200721-refactor-restore-for-rebuild-enabling.md:33]: Does this mean wait till state healthy and only then will the volume be available as long as there is at least one healthy replica?

File: enhancements/20200721-refactor-restore-for-rebuild-enabling.md [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+
+#### Replica crash on restore volume
+1. Users create a restore volume and wait for restore complete.
+2. When the restore is in progress, some replicas somehow get crashed. Then the volume rebuilds new replicas immediately, and it will become `Healthy` once the new replicas start rebuilding.
+4. The volume will be detached automatically once the restore and the rebuild complete.
+
+#### Replica crash on DR volume
+1. Users create a DR volume.
+2. Some replicas get crashed. Then the DR volume automatically rebuilds new replicas and restores the latest backup for the rebuild replicas.
+3. Users try to activate the DR volume. The DR volume will wait for the rebuild and laest backup restore complete before detachment.
```
Comment (joshimoo) [inline enhancements/20200721-refactor-restore-for-rebuild-enabling.md:45]:
```suggestion
3. Users try to activate the DR volume. The DR volume will wait for the rebuild of all replicas and successfull restoration of the latest backup before detachment.
```

File: enhancements/20200721-refactor-restore-for-rebuild-enabling.md (hunk 3 above)
Comment (shuo-wu) [reply enhancements/20200721-refactor-restore-for-rebuild-enabling.md:21]: `the last backup` means the most recently restored backup.
`the latest backup` does means the newest backup and it hasn't been restored by the volume.

I will add more explanations for this part.

File: enhancements/20200721-refactor-restore-for-rebuild-enabling.md (hunk 4 above)
Comment (shuo-wu) [reply enhancements/20200721-refactor-restore-for-rebuild-enabling.md:33]: NO. The restore/DR volume will be `Healthy` as long as all replicas are running, which is exactly the same as normal volumes.

Previously, Longhorn won't replenish replicas, and restore/DR volumes will keep state `Degraded` if there is a replica crashing.

**Comment by shuo-wu** [comment]
Updated the 2nd part of the proposal. Fixed some typos and syntax errors.

**Comment by shuo-wu** [comment]
Since the refactor is complicated, I cannot consider all details for this issue and write them to the doc. As a result, this LEP doc will be updated after the implementation. It's better not to merge it for now.

**Comment by yasker** [comment]
@shuo-wu Don't worry about that. We can update after merging it if needed.

**Comment by shuo-wu** [comment]
@yasker @joshimoo Is this good to go?

**Pull request merged**